### PR TITLE
Dev/nom7 smb v0.1

### PR DESCRIPTION
**Pull request by chifflier**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4165

Describe changes:
- Followup work from #6589 
- Rewrite SMB 2/3/4 parsers to nom 7
- As previously, no rustfmt, only affected files are changed
- All parsers but a tiny number are updated. Some dependencies (like x509-parser, RDP parsing etc.) are breaking CI on ubuntu 16.04 when upgraded, this needs to be investigated separately.
- This PR includes all changes that can be done on SMB without changing version requirements or dependencies

### CODE REVIEW

File: rust/src/common.rs [hunk 1]
```diff
@@ -1,6 +1,23 @@
 use std::ffi::CString;
 use std::os::raw::c_char;
 
+pub mod nom7 {
+    use nom7::bytes::streaming::{tag, take_until};
+    use nom7::error::ParseError;
+    use nom7::IResult;
+
+    pub fn take_until_and_consume<'a, E: ParseError<&'a [u8]>>(t: &'a [u8])
+         -> impl Fn(&'a [u8]) -> IResult<&'a [u8], &'a [u8], E>
```
Comment (jasonish) [inline rust/src/common.rs:10]: I think this needs a comment. It wasn't clear to me that it consumes that terminator until I read the code.  My first that was "take_until" does consume...

**Review by jasonish — COMMENTED** [review]
Looks great on first pass.

File: rust/src/common.rs [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+    {
+        move |i: &'a [u8]| {
+            let t = t.clone();
```
Comment (jasonish) [inline rust/src/common.rs:13]: I don't think the clone is required here (hint from clippy).

File: rust/src/common.rs (hunk 1 above)
Comment (chifflier) [reply rust/src/common.rs:10]: Ok, I'll add a comment to the function documentation

File: rust/src/common.rs (hunk 2 above)
Comment (chifflier) [reply rust/src/common.rs:13]: Indeed, this was a leftover from taking the nom combinator, and specializing it to `&[u8]`.
Fixed!

**Comment by chifflier** [comment]
Replaced by #6667

**Pull request closed without merging**